### PR TITLE
ASoC: SOF: intel: hda-stream: Print stream name on STREAM_SD_OFFSET t…

### PR DESCRIPTION
**Pull request by ujfalusi**
…imeout

In order to provide more information in case of timeout observed while
reading STREAM_SD_OFFSET, print out the stream name or in case there is
no audio stream associated (like dma-trace), print "--"

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

### CODE REVIEW

**Comment by ujfalusi** [comment]
@plbossart, @lgirdwood: FYI. Is this something can be of use as mentioned in https://github.com/thesofproject/sof/pull/4548?

File: sound/soc/sof/intel/hda-stream.c [hunk 1]
```diff
@@ -307,27 +302,29 @@ int hda_dsp_stream_trigger(struct snd_sof_dev *sdev,
...
 	default:
 		dev_err(sdev->dev, "error: unknown command: %d\n", cmd);
 		return -EINVAL;
 	}
 
-	return 0;
+	if (ret < 0) {
+		dev_err(sdev->dev,
+			"%s: cmd %d on substream %s: timeout on STREAM_SD_OFFSET read\n",
+			__func__, cmd, hstream->substream ? hstream->substream->name : "--");
```
Comment (lgirdwood) [inline sound/soc/sof/intel/hda-stream.c]: Would it be correct to assume that only the trace and or probes would not have a substream name ? if so, could we asign then a name as part of their open() ?

Comment (ujfalusi) [reply sound/soc/sof/intel/hda-stream.c]: neither the probes, dma-trace or the real compressed would have substream. We don't have place to assign custom name, let me try something...

Comment (ujfalusi) [reply sound/soc/sof/intel/hda-stream.c]: We could print out the `stream_tag` which is always unique

**Comment by ujfalusi** [comment]
Changes since v1:
- helper function to get the stream name (`hda_hstream_get_stream_name`)
 - return either the substream->name or the cstream->name or NULL

File: sound/soc/sof/intel/hda-stream.c (hunk 1 above)
Comment (lgirdwood) [reply sound/soc/sof/intel/hda-stream.c]: ok, ```stream_tag``` is in the FW logs so could be aligned with the dmesg and would work for debug.

Comment (ujfalusi) [reply sound/soc/sof/intel/hda-stream.c]: Let me try to squeeze it there. We are 90 char already which is the limit I rather not step, but under 100 is still OK for checkpatch

**Comment by ujfalusi** [comment]
Changes since v2:
- print the stream_tag as well as it is unique for each stream, regardless of type (audio/compressed/dtrace)
- replace `substream %s` with `stream %s` in the prints to save three characters, now the longest line is at 92 char

**Comment by ujfalusi** [comment]
@ranj063,
dma-trace: `hda_dsp_stream_trigger: cmd 1 on stream -- (1): timeout on STREAM_SD_OFFSET read`
aplay -Dplughw:0,0: `hda_dsp_stream_trigger: cmd 1 on stream subdevice #0 (1): timeout on STREAM_SD_OFFSET read`

on a tgl-h laptop.

**Comment by ranj063** [comment]
> @ranj063,
> dma-trace: `hda_dsp_stream_trigger: cmd 1 on stream -- (1): timeout on STREAM_SD_OFFSET read`
> aplay -Dplughw:0,0: `hda_dsp_stream_trigger: cmd 1 on stream subdevice #0 (1): timeout on STREAM_SD_OFFSET read`
> 
> on a tgl-h laptop.

@ujfalusi Thanks. I wonder if it then makes sense to also print the direction? subdevice #0 ?

**Comment by plbossart** [comment]
> > @ranj063,
> > dma-trace: `hda_dsp_stream_trigger: cmd 1 on stream -- (1): timeout on STREAM_SD_OFFSET read`
> > aplay -Dplughw:0,0: `hda_dsp_stream_trigger: cmd 1 on stream subdevice #0 (1): timeout on STREAM_SD_OFFSET read`
> > on a tgl-h laptop.
> 
> @ujfalusi Thanks. I wonder if it then makes sense to also print the direction? subdevice #0 ?

I tried to use the stream name for SoundWire stuff and all I could get was a bunch of 'subdevice #0' and 'subdevice #1'. Not very helpful when you have multiple dailinks. I ended-up adding the dai name to at least know which stream this really was.

**Comment by ujfalusi** [comment]
> > > @ranj063,
> > > dma-trace: `hda_dsp_stream_trigger: cmd 1 on stream -- (1): timeout on STREAM_SD_OFFSET read`
> > > aplay -Dplughw:0,0: `hda_dsp_stream_trigger: cmd 1 on stream subdevice #0 (1): timeout on STREAM_SD_OFFSET read`
> > > on a tgl-h laptop.
> > 
> > 
> > @ujfalusi Thanks. I wonder if it then makes sense to also print the direction? subdevice #0 ?
> 
> I tried to use the stream name for SoundWire stuff and all I could get was a bunch of 'subdevice #0' and 'subdevice #1'. Not very helpful when you have multiple dailinks. I ended-up adding the dai name to at least know which stream this really was.

@plbossart, @ranj063, or if we can reach the dai-link name for the given stream? Looking at the prints the `stream_tag` is not the most useful information I have seen, but when multiple streams are running they should give some pointer.
I'll check tomorrow if we can swap the substream->name with something more helpful.

**Comment by ujfalusi** [comment]
Changes since v3:
- the printed information contains more and more accurate information

Tested on top of SOF clients branch to have working probes and the prints are:
**dma-trace**
`hda_dsp_stream_trigger: cmd 1 on -- (Capture, stream_tag: 1): timeout on STREAM_SD_OFFSET read`

**aplay -Dplughw:1,0**
`hda_dsp_stream_trigger: cmd 1 on dai_link "HDA Analog" (Playback, stream_tag: 1): timeout on STREAM_SD_OFFSET read`

**probes crecord**
`hda_dsp_stream_trigger: cmd 1 on dai_link "Compress Probe Capture" (Capture, stream_tag: 2): timeout on STREAM_SD_OFFSET read`

File: sound/soc/sof/intel/hda-stream.c [hunk 2]
```diff
@@ -25,6 +25,35 @@
...
+{
+	struct snd_soc_dai_link *dai_link;
+	struct snd_soc_pcm_runtime *rtd;
+
+	if (hstream->substream)
+		rtd = asoc_substream_to_rtd(hstream->substream);
+	else if (hstream->cstream)
+		rtd = hstream->cstream->private_data;
+	else
+		return kasprintf(GFP_KERNEL, "-- (%s, stream_tag: %u)",
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-stream.c:46]: for my education, when would we hit the last else statement?

Comment (ujfalusi) [reply sound/soc/sof/intel/hda-stream.c:46]: dma-trace is not audio, we don't have substream nor cstream. We could print `"dma-trace"' which would be correct at the moment, afaik we only hijack the DMA from audio for that purpose only.

Comment (plbossart) [reply sound/soc/sof/intel/hda-stream.c:46]: please add a comment then, this is confusing and not obvious for reviewers. Thanks!

Comment (ujfalusi) [reply sound/soc/sof/intel/hda-stream.c:46]: Done

**Comment by ujfalusi** [comment]
Changes since v4:
- Add comment for non audio user case in `hda_hstream_dbg_get_stream_info_str()`
- remove the `dai_link` casting, simply use `rtd->dai_link->name`

**Pull request merged**